[Ivan Wong] iP

wpinrui opened the pull request:
# Good Janet :hugs:
> Elenor told me that instead of being sad, I should, "Go get it, girl." So I'm going to go get it, girl. I'll get everything, just to be safe. - Good Janet [(source)](https://img.buzzfeed.com/buzzfeed-static/static/2019-09/5/18/asset/38945558c66c/sub-buzz-283-1567708506-1.jpg?downsize=600:*&output-format=auto&output-quality=auto)

Good Janet helps you track your tasks. Here are some features:
- GUI
- Easy to learn
- **Super fast to use!**

## Getting started
1. Download the **GUI version** of the JAR file in the releases
2. Open a terminal window in the JAR file location
3. Run the following: `java -jar janet.jar`

## Planned features
- [x] Add tasks
- [x] Delete tasks
- [ ] Update tasks

## Advanced Users
The save file is currently hardcoded. If you want to change where the save file is located, you can edit the source code and recompile the JAR file. Simply change "janet.txt" in the constructor to any custom path you like, e.g. "C:/Users/User/Documents/janet.txt" and it should work *as long as the folder exists*. 

```java
public Janet() {
        storage = new Storage("janet.txt"); // this path can be changed
        try {
            tasks = storage.readSave();
        } catch (IOException | ClassNotFoundException e) {
            System.out.println(e);
        }
    }
```

Commit:
Add Gradle support

wpinrui committed:
Ignore .class files

wpinrui committed:
Implement skeletal version of Good Duke

wpinrui committed:
Add ability to store text and list it

wpinrui committed:
Add Task class

wpinrui committed:
Task: Add toString method

wpinrui committed:
Add Mark as Done

wpinrui committed:
Add classes for Deadline, Event, ToDos

wpinrui committed:
Add support for ToDos, Events, Deadlines

wpinrui committed:
Add testcase for todo, deadline, event, list, done, bye

wpinrui committed:
Add expected result of testcase

wpinrui committed:
Add Command class to handle user input

wpinrui committed:
Add exception handling for user input via DukeException

wpinrui committed:
Replace Task[] tasklist with ArrayList<Task> tasklist

wpinrui committed:
Add ability to delete tasks

wpinrui committed:
Reindent switch case to follow style guide

wpinrui committed:
Add save and load save

wpinrui committed:
Import LocalDate

wpinrui committed:
Merge branch 'branch-Level-7'

wpinrui committed:
Add support for date entry for event / deadline

wpinrui committed:
Merge branch 'branch-Level-8'

wpinrui committed:
Added Ui class and made static fields non-static

wpinrui committed:
Add storage class

wpinrui committed:
Add Parser class to interpret user input

wpinrui committed:
Add TaskList class

wpinrui committed:
Package Java files into duke

wpinrui committed:
Add tests for Deadline and TaskList classes

wpinrui committed:
Add metadata for building JAR

wpinrui committed:
Move JUnit tests to package dukeTest

wpinrui committed:
Add JavaDoc to public methods and classes

wpinrui committed:
Edit style to comply with coding standard

wpinrui committed:
Add find operation to filter tasks by query

wpinrui committed:
Edit style according to Java conventions

wpinrui committed:
Add find operation

wpinrui committed:
Merge branch 'branch-A-JavaDoc'

wpinrui committed:
Merge branch 'master' of https://github.com/wpinrui/ip

wpinrui committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Event.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/Ui.java

wpinrui committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Ui.java

tzejit commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,98 @@
+     *
+     * @return Duke operation
+     */
+    public String getOperation() {
+        return operation;
+    }
```
I like how you used getters to get information instead of accessing it directly from the instance itself!

tzejit commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,77 @@
+     * @throws DukeException If invalid or insufficient arguments are provided
+     */
+    public static Command parseUserInput(String input) throws DukeException {
+        String[] strings = input.split(" ");
```
Should this variable have a more intuitive name?

wpinrui replied:
Thanks for spotting it. I've changed it to `inputStringParts`.

tzejit commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,66 @@
+     *
+     * @return Number of tasks in list
+     */
+    public int size() {
```
Maybe the method can be named after a verb more such as getSize?

wpinrui replied:
I'm unsure about this, because the method for ArrayList is named 'size'.

tzejit reviewed (commented):
Overall really good code quality! Almost all namings follow the proper convention, good OOP practices are followed and methods have well written javadoc comments!

wpinrui committed:
Merge branch 'add-gradle-support' of https://github.com/wpinrui/ip

wpinrui committed:
Add Gradle checkstyle files

wpinrui committed:
Fix style issues from running style checks

wpinrui committed:
Add placeholder GUI

wpinrui committed:
Add user input response to GUI

wpinrui committed:
Restore CLI functionality to GUI

wpinrui committed:
Add FXML files for GUI

wpinrui committed:
Improve GUI styling with colour and padding

wpinrui committed:
Merge branch 'branch-Level-10'

wpinrui committed:
Fix bug where catch block was causing exception to be thrown

wpinrui committed:
Fix issue where message was cut off due to height constraint

wpinrui committed:
Fix duplicated text display during find command

wpinrui committed:
Change bot's name to Janet

wpinrui committed:
Merge branch 'rename-to-janet'

wpinrui committed:
Parser: change strings to a more intuitive name

wpinrui committed:
Update build.gradle to reflect project structure

wpinrui committed:
Add assertions

wpinrui committed:
Improve code quality through refactoring

The code base has issues like long methods, excessive nesting
and long switch cases.

A refactored codebase will be better for expandability and bug
testing because it is more readable and it is easy to find out
where the problematic code lies.

Let's,
* extract the bodies of switch cases into their own private methods
so that each case can be one or two lines long
* make use of guard clauses to make the main path more prominent
* practise using single level of abstraction
* use named constants to eliminate magic numbers
* use logical names to communicate ideas better

We follow the guidelines listed here to improve readability:
https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html

wpinrui committed:
Merge pull request #1 from wpinrui/branch-A-Assertions

Add assertions

wpinrui committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/janet/Deadline.java
#	src/main/java/janet/Event.java
#	src/main/java/janet/Janet.java
#	src/main/java/janet/MainWindow.java

wpinrui committed:
Merge pull request #2 from wpinrui/branch-A-CodeQuality

Branch a code quality

wpinrui committed:
Add schedule command to view tasks for a certain date

wpinrui committed:
Merge branch 'branch-B-ViewSchedules'

wpinrui committed:
TaskList: resolve style issues

wpinrui committed:
Replace Janet and Eleanor profile images with PNGs

The JPG versions of Janet and Eleanor do not have transparent
backgrounds.

This means that the chat messages displayed are less visually
appealing.

Let's replace them with PNGs in which their faces are cropped in
a circle and the rest of the image is transparent.

Using circular images is associated with chat boxes and lets the
user feel like they are really chatting with a real person.

itsraveen commented on src/main/java/janet/Deadline.java:
```diff
@@ -0,0 +1,57 @@
+
+    /**
+     * Returns a formatted String representing the deadline of the task in the form
+     * MMM d YYYY.
```
I like how your JavaDoc here is very clear with what the method does. I think its great that you included 'MMM d YYYY' in your code.

itsraveen commented on src/main/java/janet/Janet.java:
```diff
@@ -0,0 +1,126 @@
+        default:
+            return "";
+        }
+    }
```
Code very well written with each of the events being handles very well. It becomes clear when that you know what each of the cases does since you have a method to handle each cases.

itsraveen commented on src/main/java/janet/Janet.java:
```diff
@@ -0,0 +1,126 @@
+        return Ui.doneString(task);
+    }
+
+    private String handleNewToDo(Command command) throws IOException {
+        ToDo task = new ToDo(command.getDescription());
+        tasks.add(task);
+        storage.writeSave(tasks);
+        return Ui.addedString(task, tasks.size());
+    }
+
+    private String handleNewDeadline(Command command) throws IOException {
+        Deadline task = new Deadline(command.getDescription(), command.getTime());
+        tasks.add(task);
+        storage.writeSave(tasks);
+        return Ui.addedString(task, tasks.size());
+    }
+
+    private String handleNewEvent(Command command) throws IOException {
+        Event task = new Event(command.getDescription(), command.getTime());
+        tasks.add(task);
+        storage.writeSave(tasks);
+        return Ui.addedString(task, tasks.size());
+    }
```
Would this be better if it was all in a single method and handles the different cases? Most of the code are repeated code so it might help to make it shorter. Other than that, looks good to me.

wpinrui replied:
Thanks for your feedback. I was considering doing this but realised I would have many if-else statements. Now that I think about it further, it may be possible to use polymorphism to achieve this 😄

itsraveen commented on src/main/java/janet/Parser.java:
```diff
@@ -0,0 +1,142 @@
+        String[] inputStringParts = input.split(" ");
+        String operation = inputStringParts[0];
+        assert(inputStringParts[0] != null);
+        switch (operation) {
+        case "bye":
+            // Fallthrough
+        case "list":
+            return new Command(operation);
+        case "done":
+            return parseDone(inputStringParts);
+        case "todo":
+            return parseTodo(input, inputStringParts);
+        case "deadline":
+            return parseDeadline(input, inputStringParts);
+        case "event":
+            return parseEvent(input, inputStringParts);
+        case "delete":
+            return parseDelete(inputStringParts);
+        case "find":
+            return parseFind(input, inputStringParts);
+        case "schedule":
+            return parseSchedule(input, inputStringParts);
+        default:
+            throw new JanetException(Ui.UNRECOGNISED_OPERATION);
```
Once again the switch statements are handled very well!

itsraveen reviewed (commented):
Code looks clean with all the JavaDocs being clear and concise. Looks good to me!

Linda124 commented on src/main/java/janet/Ui.java:
```diff
@@ -0,0 +1,97 @@
+     * @param taskList Janet's list of tasks
+     * @return String representation of list
+     */
+    public static String taskListString(TaskList taskList) {
```
Perhaps you could name this method as printTaskListString so that it sounds more like a verb?

wpinrui replied:
Ahh, good catch. Thank you!

Linda124 commented on src/main/java/janet/Ui.java:
```diff
@@ -0,0 +1,97 @@
+     * @param index The index of the non-existent task that the user tried to access
+     * @return String to inform the user of index out of bounds error
+     */
+    public static String outOfBoundsString(int index) {
```
Would printOutOfBoundsString be a bit more intuitive here? I noticed that a similar approach could be used for the methods "deletedString", "doneString" and "addedString".

wpinrui replied:
I'm a little unsure on this, because this method returns a String whereas printXXX might give the impression of a void method that has the side-effect of logging to console. Thanks for pointing it out, I may open an issue to ask for more opinions.

Linda124 commented on src/main/java/janet/TaskList.java:
```diff
@@ -0,0 +1,100 @@
+     *
+     * @return Number of tasks in list
+     */
+    public int size() {
```
Would it be better to name this method getSize so that the function of the method is better understood?

Linda124 commented on src/main/java/janet/TaskList.java:
```diff
@@ -0,0 +1,100 @@
+     * @param index Index of the task in the list of tasks
+     * @return Task found at that index in the list
+     */
+    public Task get(int index) {
```
Since I suggested that size could be changed to getSize, perhaps you could consider changing this method name to getIndex as well.

Linda124 commented on src/main/java/janet/Janet.java:
```diff
@@ -0,0 +1,126 @@
+        return Ui.doneString(task);
+    }
+
+    private String handleNewToDo(Command command) throws IOException {
```
I like how most of your method names and variable names are easy to understand.

Linda124 reviewed (commented):
Overall, the code quality is really good! Generally, good naming conventions were used.

wpinrui committed:
Add acknowledgements for reused code

wpinrui committed:
Ui: change names to sound like actions

wpinrui committed:
Set theme jekyll-theme-slate

wpinrui committed:
Add screenshot for user guide

wpinrui committed:
Create User Guide

wpinrui committed:
Remove unsupported emoji from user guide

wpinrui committed:
Update README.md

wpinrui committed:
Update tests to follow coding standard

wpinrui committed:
Merge branch 'master' of https://github.com/wpinrui/ip

wpinrui committed:
Fix a bug with entering an index for done that does not exist

wpinrui committed:
Fix typo in readme and user guide

wpinrui committed:
Update README.md
